File: (结构)代理模式.v2/php/gpt/index.代理.php
<?php
namespace php\gpt\proxy;

//Subject接口：定义了RealSubject和Proxy的共同接口，这样在任何使用RealSubject的地方都可以使用Proxy。
//RealSubject类：定义了代理所代表的实际对象。
//Proxy类：维护一个对RealSubject的引用，控制对RealSubject的访问，并可以负责创建和删除它。
//Client：使用Subject接口执行它的任务，无需关心工作是由RealSubject完成还是通过Proxy完成。

interface SubjectInterface {
    public function request();
}

class RealSubject implements SubjectInterface {
    public function request() {
        echo "RealSubject: Handling request.\n";
    }
}

class Proxy implements SubjectInterface {
    private $realSubject;

    public function __construct(RealSubject $realSubject) {
        $this->realSubject = $realSubject;
    }

    public function request() {
        if ($this->checkAccess()) {
            $this->realSubject->request();
            $this->logAccess();
        }
    }

    private function checkAccess() {
        // 检查访问权限
        echo "Proxy: Checking access prior to firing a real request.\n";
        return true; // 假设所有的客户端访问都是被允许的
    }

    private function logAccess() {
        // 记录请求日志
        echo "Proxy: Logging the time of request.\n";
    }
}

// Client code
echo "Client: Executing the client code with a real subject:\n";
$realSubject = new RealSubject();
$realSubject->request();

echo "\n";

echo "Client: Executing the same client code with a proxy:\n";
$proxy = new Proxy($realSubject);
$proxy->request();

//
//    Client: Executing the client code with a real subject:
//    RealSubject: Handling request.
//
//    Client: Executing the same client code with a proxy:
//    Proxy: Checking access prior to firing a real request.
//    RealSubject: Handling request.
//    Proxy: Logging the time of request.

File: (结构)02.代理模式.v2/golang/gpt/index.代理.go
package main

import "fmt"

type Subject interface {
	DoAction()
}

type RealSubject struct{}

func (rs *RealSubject) DoAction() {
	fmt.Println("RealSubject: Doing action")
}

type Proxy struct {
	realSubject *RealSubject
}

func (p *Proxy) DoAction() {
	if p.checkAccess() {
		p.realSubject.DoAction()
		p.logAccess()
	}
}

// 预留的钩子方法, 在这里可以进行权限检查
func (p *Proxy) checkAccess() bool {
	fmt.Println("Proxy: Checking access before firing a real request.")
	// 这里应该包含一些真实的检查逻辑
	// 如果合适，返回true
	return true
}

// 可以记录每次的请求
func (p *Proxy) logAccess() {
	fmt.Println("Proxy: Logging the time of request.")
}

/**
在这个例子中，main函数创建了一个真实对象realSubject，
	然后创建了代理对象proxy并传入了realSubject作为其依赖。
	当调用proxy.DoAction()时，proxy会先执行权限检查和日志记录，然后调用真实对象的DoAction()方法。

使用代理模式有许多好处，你可以不修改客户代码的基础上增加或修改对象的行为，
	可以非常方便地实现懒加载、权限检查、日志记录等功能，并且能够更好地遵循开闭原则。
*/
func main() {
	realSubject := &RealSubject{}
	proxy := Proxy{realSubject: realSubject}

	// 客户端使用代理，而不是直接调用真实对象
	proxy.DoAction()
	/**
	Proxy: Checking access before firing a real request.
	RealSubject: Doing action
	Proxy: Logging the time of request.
	*/
}
